### ISSUE TITLE
AddressSanitizer: heap-use-after-free in xname_flags 

### ISSUE THREAD

**Issue by elunna** [elunna/hackem#95]
```
=================================================================
==2019==ERROR: AddressSanitizer: heap-use-after-free on address 0x60c00115631e at pc 0x555555f17eee bp 0x7fffffffd4d0 sp 0x7fffffffd4c0
READ of size 2 at 0x60c00115631e thread T0
    #0 0x555555f17eed in xname_flags /home/lunatunez/hackem/src/objnam.c:603
    #1 0x555555f17dd1 in xname /home/lunatunez/hackem/src/objnam.c:581
    #2 0x555555f03314 in mmake_wish /home/lunatunez/hackem/src/muse.c:4790
    #3 0x555555ea4785 in precheck /home/lunatunez/hackem/src/muse.c:166
    #4 0x555555edf6dd in use_misc /home/lunatunez/hackem/src/muse.c:3264
    #5 0x555555e52a43 in dochug /home/lunatunez/hackem/src/monmove.c:660
    #6 0x555555e45f89 in dochugw /home/lunatunez/hackem/src/monmove.c:122
    #7 0x555555db242b in movemon /home/lunatunez/hackem/src/mon.c:1532
    #8 0x555555766228 in moveloop /home/lunatunez/hackem/src/allmain.c:244
    #9 0x5555563f591f in main ../sys/unix/unixmain.c:353
    #10 0x7ffff7365082 in __libc_start_main ../csu/libc-start.c:308
    #11 0x555555763acd in _start (/home/lunatunez/games/hackemdir/hackem+0x20facd)

0x60c00115631e is located 30 bytes inside of 120-byte region [0x60c001156300,0x60c001156378)
freed by thread T0 here:
    #0 0x7ffff769940f in __interceptor_free ../../../../src/libsanitizer/asan/asan_malloc_linux.cc:122
    #1 0x555555d701b4 in dealloc_obj /home/lunatunez/hackem/src/mkobj.c:2497
    #2 0x5555560e3259 in obfree /home/lunatunez/hackem/src/shk.c:1162
    #3 0x555555b0a636 in merged /home/lunatunez/hackem/src/invent.c:806
    #4 0x555555d6f232 in add_to_minv /home/lunatunez/hackem/src/mkobj.c:2375
    #5 0x5555561a9025 in mpickobj /home/lunatunez/hackem/src/steal.c:570
    #6 0x555555f01d5a in mmake_wish /home/lunatunez/hackem/src/muse.c:4779
    #7 0x555555ea4785 in precheck /home/lunatunez/hackem/src/muse.c:166
    #8 0x555555edf6dd in use_misc /home/lunatunez/hackem/src/muse.c:3264
    #9 0x555555e52a43 in dochug /home/lunatunez/hackem/src/monmove.c:660
    #10 0x555555e45f89 in dochugw /home/lunatunez/hackem/src/monmove.c:122
    #11 0x555555db242b in movemon /home/lunatunez/hackem/src/mon.c:1532
    #12 0x555555766228 in moveloop /home/lunatunez/hackem/src/allmain.c:244
    #13 0x5555563f591f in main ../sys/unix/unixmain.c:353
    #14 0x7ffff7365082 in __libc_start_main ../csu/libc-start.c:308

previously allocated by thread T0 here:
    #0 0x7ffff7699808 in __interceptor_malloc ../../../../src/libsanitizer/asan/asan_malloc_linux.cc:144
    #1 0x5555557700bb in alloc /home/lunatunez/hackem/src/alloc.c:46
    #2 0x555555d638b6 in mksobj /home/lunatunez/hackem/src/mkobj.c:806
    #3 0x555555f01d12 in mmake_wish /home/lunatunez/hackem/src/muse.c:4777
    #4 0x555555ea4785 in precheck /home/lunatunez/hackem/src/muse.c:166
    #5 0x555555edf6dd in use_misc /home/lunatunez/hackem/src/muse.c:3264
    #6 0x555555e52a43 in dochug /home/lunatunez/hackem/src/monmove.c:660
    #7 0x555555e45f89 in dochugw /home/lunatunez/hackem/src/monmove.c:122
    #8 0x555555db242b in movemon /home/lunatunez/hackem/src/mon.c:1532
    #9 0x555555766228 in moveloop /home/lunatunez/hackem/src/allmain.c:244
    #10 0x5555563f591f in main ../sys/unix/unixmain.c:353
    #11 0x7ffff7365082 in __libc_start_main ../csu/libc-start.c:308

SUMMARY: AddressSanitizer: heap-use-after-free /home/lunatunez/hackem/src/objnam.c:603 in xname_flags
Shadow bytes around the buggy address:
  0x0c1880222c10: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c1880222c20: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c1880222c30: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c1880222c40: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c1880222c50: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
=>0x0c1880222c60: fd fd fd[fd]fd fd fd fd fd fd fd fd fd fd fd fa
  0x0c1880222c70: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c1880222c80: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c1880222c90: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c1880222ca0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c1880222cb0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07 
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
  Shadow gap:              cc
==2019==ABORTING

```

**Comment by elunna** [comment]
Monster got a wish from a lamp.
The item possibly merged before getting to xname: cockatrice egg, potion, etc.

**Comment by elunna** [comment]
Fixed in commit 0eb9fd68a
